docs/architetture-software/README.md

```diff
@@ -0,0 +1,212 @@
+
+## Microkernel
+
+Chiamata anche _architettura a plugin_, si compone essenzialmente di un nucleo centrale che si occupa di gestire le comunicazioni tra i vari componenti, che sono sviluppati e distribuiti come plugin.
+Questi plugin possono essere sviluppati e distribuiti in maniera indipendente, ma funzionando attorno ad un nucleo centrale, faranno sempre parte dello stesso _blocco_, inteso che l'applicativo online è unico, come nel caso del monolite _classico_.
+
+Il concetto principale attorno a questa architettura è l'indipendenza totale tra i plugin. Per definizione, nessun plugin dovrebbe avere come dipendenza un altro plugin.
+
+In questa architettura il database, il back-end e il front-end sono tutti parte dello stesso processo/applicativo.
+
+- **Start-up**: 4. Essendo una variante del monolite, lo start-up è molto semplice, ma la necessità di organizzare il codice in un nucleo centrale e implementare un sistema di plugin può richiedere più tempo.
+- **Costi**: 3. Pur essendo un'architettura _a costo zero_ in termini di start-up, il costo cresce esponenzialmente per via della necessità di _scalare orizzontalmente_.
+- **Sviluppo e manutenzione**: 4. Trattasi di un'architettura semplice da mantenere e sul quale aggiungere funzionalità, ma la necessità di non far comunicare i plugin tra di loro può aggiungere una lieve complessità.
+- **Tracciabilità**: 3. Testare e tracciare eventuali bug in questa architettura è mediamente complesso. L'isolamento dei plugin aiuta con il testing, ma la correlazione col nucleo centrale a volte può rallentare il processo di tracciamento.
+- **Scalabilità**: 2. Le difficoltà di scalabilità sono anche in questo caso dovute alla _singola unità di rilascio_ o _single deployment unit_, che non ne permette la _scalabilità orizzontale_.
+- **Performance**: 4. Processi di comunicazione semplici e isolabilità rendono questa architettura molto performante.
+
+## Microservizi
+
+La _buzzword_ per eccellenza degli ultimi anni.
+
+Spesso affiancata alle architetture orientate agli eventi, questa architettura software si compone di un insieme di servizi indipendenti tra di loro, ognuno con un compito ben definito, che comunicano tra di loro attraverso un meccanismo di comunicazione _solitamente_ asincrono.
+
+A differenza delle architetture _a singolo blocco_ viste sopra, questa architettura permette di scalare i singoli servizi, in quanto ogni servizio viene rilasciato in maniera indipendente.
+
+In questa architettura ogni microservizio solitamente ha il proprio database e il proprio back-end (basato su API), mentre il front-end è solitamente un'interfaccia grafica che comunica con i microservizi attraverso le API, a prescindere da dove risieda o da come sia implementato. I database sono solitamente piccoli e si occupano di gestire solo i dati necessari al microservizio in questione.
```

Comment:
sarei molto categorico a riguardo (per il bene del lettore). Ogni ms ha il suo db e ci puoi accedere solo tramite le api del ms. citando full metal jacket: Questo è il mio database. Ce ne sono tanti come lui ma questo è il mio.

Reply:
@frnmjn dipende dal grado di opinione che vogliamo dare.
Statisticamente è giusto dire che _solitamente_ ogni ms ha il proprio servizio, mentre dire che **deve** averlo è un po' forte in quanto nessuno vieta di farlo.

Reply:
Ok, già così mi piace di più.
Magari aggiungo un punto a riguardo, ma "puoi" e "devi" hanno tantissime sfumature in mezzo.

Reply:
Ho aggiunto questo paragrafo, cosa ne pensi?

`A differenza delle altre architetture citate, la condivisione del database è da considerarsi un _anti-pattern_, ovvero una pratica da non adottare, in quanto può portare a problemi di _coupling_ tra i microservizi, esattamente l'opposto di quanto si vuole ottenere con questa architettura.`

Reply:
mi piace!